pymc3/distributions/distribution.py

```diff
@@ -602,160 +653,88 @@ def generate_samples(generator, *args, **kwargs):
-                if samples[0].shape == broadcast_shape:
-                    samples = samples[0]
-                else:
-                    suffix = broadcast_shape[len(size_tup):] + dist_shape
-                    samples.extend([generator(*args, **kwargs).
-                                    reshape(broadcast_shape)[..., np.newaxis]
-                                    for _ in range(np.prod(suffix,
-                                                           dtype=int) - 1)])
-                    samples = np.hstack(samples).reshape(size_tup + suffix)
-            else:
-                # The parameter shape is given, but we have to concatenate it
-                # with the size tuple
-                samples = generator(size=size_tup + broadcast_shape,
-                                    *args,
-                                    **kwargs)
-        else:
-            samples = None
-    # Args have been broadcast correctly, can just ask for the right shape out
-    elif dist_shape[-len(broadcast_shape):] == broadcast_shape:
-        samples = generator(size=size_tup + dist_shape, *args, **kwargs)
-    # Inputs have the right size, have to manually broadcast to the right dist_shape
-    elif broadcast_shape[:len(size_tup)] == size_tup:
-        suffix = broadcast_shape[len(size_tup):] + dist_shape
-        samples = [generator(*args, **kwargs).reshape(size_tup + (1,)) for _ in range(np.prod(suffix, dtype=int))]
-        samples = np.hstack(samples).reshape(size_tup + suffix)
-    else:
-        samples = None
 
-    if samples is None:
+    try:
+        dist_bcast_shape = broadcast_dist_samples_shape(
+            [_dist_shape, broadcast_shape],
+            size=size,
+        )
+    except (ValueError, TypeError):
         raise TypeError('''Attempted to generate values with incompatible shapes:
             size: {size}
             size_tup: {size_tup}
             broadcast_shape[:len(size_tup)] == size_tup: {test}
             dist_shape: {dist_shape}
             broadcast_shape: {broadcast_shape}
-        '''.format(size=size, size_tup=size_tup, dist_shape=dist_shape, broadcast_shape=broadcast_shape, test=broadcast_shape[:len(size_tup)] == size_tup))
+        '''.format(size=size,
+                   size_tup=size_tup,
+                   dist_shape=dist_shape,
+                   broadcast_shape=broadcast_shape,
+                   test=broadcast_shape[:len(size_tup)] == size_tup)
+        )
+    if dist_bcast_shape[:len(size_tup)] == size_tup:
+        samples = generator(size=dist_bcast_shape, *args, **kwargs)
+    else:
+        samples = generator(size=size_tup + dist_bcast_shape, *args, **kwargs)
+    samples = np.asarray(samples)
 
     # reshape samples here
-    if samples.shape[0] == 1 and size == 1:
-        if len(samples.shape) > len(dist_shape) and samples.shape[-len(dist_shape):] == dist_shape:
+    if samples.ndim > 0 and samples.shape[0] == 1 and size_tup == (1,):
```

Comment:
All of the following reshape conditions are from previous versions of `generate_samples`. They effectively do the following:
1. It makes `size=1` return the same shape as `size=None`.
2. It forces RVs of `shape=1` to return samples as if they were scalars of `shape=tuple()`.

I think that these conditions should be dropped because they add ambiguity. What do you think @ColCarroll?